### ISSUE TITLE
Very strange KinovaCommException instance [Case #CS001430]

### ISSUE THREAD

**Issue by ZhengYi0310** [Kinovarobotics/kinova-ros#233]
Hi,
   I'm using [ros_control](http://wiki.ros.org/ros_control) to rewrite the ros package, in order to implement my own control loop later, and in my [hardware_interface](http://wiki.ros.org/hardware_interface?distro=indigo), I implemented [a general read and write loop](http://wiki.ros.org/ros_control/Tutorials/Create%20your%20own%20hardware%20interface) use leveraging KinovaComm, and the following is the code snippet of the hardware interface:
![1](https://user-images.githubusercontent.com/8685125/56555260-3ce26a00-6549-11e9-8489-2662b972c0d6.png)
![2](https://user-images.githubusercontent.com/8685125/56555276-45d33b80-6549-11e9-9f34-973c717d92aa.png)
![3](https://user-images.githubusercontent.com/8685125/56555283-4cfa4980-6549-11e9-810c-e59101d09b40.png)
![5](https://user-images.githubusercontent.com/8685125/56555289-51266700-6549-11e9-91bb-399c375b35c5.png)

in the following is the code snippet for the control loop:
![6](https://user-images.githubusercontent.com/8685125/56555403-ab272c80-6549-11e9-8da3-bbc55b84ac6f.png)

Then, there is another node that will call the KinovaHWInterface::setTorqueControlModeService during my experiment (imaging I want to first set the arm torque control mode, so I can give a demo to the robot, then I set the robot back to position mode, feed it a spline trajectory so it can move back to the start position), I get the following exception:
![Screenshot from 2019-04-22 21:37:05](https://user-images.githubusercontent.com/8685125/56555477-ede90480-6549-11e9-85d2-6f004715039f.png), 

And sometimes the communication exception happen several times in a row:
![8](https://user-images.githubusercontent.com/8685125/56556084-d874da00-654b-11e9-9c53-1f572a027ac7.png)

The try catch block catches a communication exception during the process, it's really strange that it only happens when I try to switch mode, I wonder what is the reason of this, because everything looks normal...The loop is running at 100hz, robot name j2n7s300, usb connection, my system is ubuntu 14.04 LTS, ros indigo, the kinova_ros package I'm using is also indigo branch.

Thanks

**Comment by sbrisebois** [comment]
Hi Yi Zheng,

My first assumption wold be that you are calling multiple API command at the same time.
The Jaco2 API is not thread safe, and using services with API call may cause big issues.

The first thing to try is to add a mutex on the kinova_comm_ object.
Each time you use the object you must gain the mutex to avoid parallel call.

Let me know if this help.

Sylvain

**Comment by ZhengYi0310** [comment]
Hi  sbrisebois,
     
Thanks for the replay !
     
When you say I'm calling multiple API command at the same time, do you mean commands like kinova_comm_->getJointAngles(), kinova_comm_->getJointVelocities(), kinova_comm_->setTorqueControlState() etc. ?
    
 If yes, then I'd like to clarify that the kinova_comm_ object is a boost::shared_ptr of kinova::KinovaComm declared in kinova-ros/kinova_driver/src/kinova_comm.cpp, which by construction, already has a internal boost::recursive mutex , and there is boost::recursive_mutex::scoped_lock lock(api_mutex_) been called inside functions like getJointAngles(), getJointVelocities(), etc.
   
 If you mean that I need to add another mutex for my control loop besides this, then it's a bit strange because I see in kinova-ros/kinova_driver/src/kinova_arm.cpp, there are also functions which will call kinova::KinovaComm function to get joint status and tool status without adding a mutex, such as KinovaArm::statusTimer(const ros::TimerEvent&).

Basically what I'm doing is just implementing my own kinova::KinovaArm class, without modifying anything in kinova::KinovaComm, but just leverage it.

Thanks

**Comment by sbrisebois** [comment]
Hi Yi Zheng,

Good catch !. Please tell us if this fix your problem, we will fix it in the repo

Best,
Sylvain

**Comment by ZhengYi0310** [comment]
Hi sbrisebois ,
    I did multiple tests, at least for my system, the problem appears to be gone.

    Thanks

**Comment by sbrisebois** [comment]
Thanks,

I will raise a flag to our eng dept to fix this in the repo.

Many thanks for your inputs and help.

Best,
Sylvain

**Comment by ZhengYi0310** [comment]
Hi sbrisebois,
    It looks like all the robot state reading functions not part of the high level api loop ? My general hardware interface loop is like this:
    
![updateloophw](https://user-images.githubusercontent.com/8685125/57101176-81060500-6cd5-11e9-841b-ae776f3115f6.png)

and my read function is like this:
![read](https://user-images.githubusercontent.com/8685125/57101244-a85cd200-6cd5-11e9-832e-58221cc82c0c.png)

and my write function is like this:
![write](https://user-images.githubusercontent.com/8685125/57101273-ba3e7500-6cd5-11e9-8338-6b56ecd981f3.png)

even if I commented out controller_manager_->update(ros::Time::now(), elapsed_time_), so bascially I'm doing nothing and all joint_velocity_command_ fields are 0, the cycle time of the loop can is 0.2 secs, as you can see in the following image:
![sample loop](https://user-images.githubusercontent.com/8685125/57101523-46509c80-6cd6-11e9-8357-52e19faa6b6c.png)

So does this basically mean we should never call functions getting robot state and functions sending command of high level API in the same thread ?

Thanks

**Comment by ZhengYi0310** [comment]
by commenting out the two lines in read function that get cartesian position and cartesian force, the cycle loop is 0.013 ~ 0.014 secs. But it's still giving me trouble, because when I uncomment the  controller_manager_->update(ros::Time::now(), elapsed_time_), I will have another thread computing desired velocities and assign the value to joint_velocity_command_, but since this loop is running lower than 100hz, the robot is unable to track the desired joint velocities.

**Comment by KinovaCustomerSupport** [comment]
Hi ZhengYi0310,

I read through your code and for me the duration of the loop is partially explained by the number of API call you do each loop.

Each API call send a USB command to the robot (it runs at 100hz, it handles each message at 10ms) and most of the command expect a return message, so each API call take several ms to run since they are blocking.

You do 6 api calls in your read alone, it could take more than 100 ms.

What I would suggest is to split the execution of the calls to read only critical information and write it back to the arm each iteration. Also, read and write the rest every once in a while (1 to 10 iteration).

Let me know if it helps,
Jean-Philippe Côté
Customer support
